xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -426,6 +496,111 @@ private String processRouteConfig(RouteConfiguration config) {
     return null;
   }
 
+  /**
+   * Handles EDS response, which contains a list of ClusterLoadAssignment messages. Each
+   * ClusterLoadAssignment message contains information for endpoints in a cluster, grouped
+   * by localities. The response is NACKed if contains invalid information for gRPC's usage.
+   * Otherwise, an ACK request is sent to management server. Information in ClusterLoadAssignment
+   * messages are saved in local cache, in case of corresponding endpoint watchers are registered
+   * later. Endpoint watchers interested in endpoint information contained in this response are
+   * notified.
+   */
+  private void handleEdsResponse(DiscoveryResponse edsResponse) {
+    logger.log(Level.FINE, "Received an EDS response: {0}", edsResponse);
+    checkState(adsStream.edsResourceNames != null,
+        "Never requested for EDS resources, management server is doing something wrong");
```

Comment:
This `checkState` has many issues:
- shouldn't it generate a NACK to server instead throwing exception?
- the `adsStream.edsResourceNames != null` check seems to specialized (special case). Ideal check would be if the any resource name from `edsResponse`  is not present in `adsStream.edsResourceNames` then "management server is doing something wrong". 
- the check should be dependent on nonce verification: the response could be to an old request in which case it could have outdated resource names.

Reply:
- This check fails only when server is sending back an EDS response while client never sent an EDS request. I consider this as violation of the protocol and should let things break immediately to raise our attention fixing it instead of letting it go with mild handling. However, on the other hand, I am also ok to allow this protocol violation and just ignore resources sent by server proactively. If server does not follow the protocol spec, client can do anything.

- EDS response can send us extra than what we requested, but those extra will be ignored. Here we are talking about the very first response. Management should not start sending responses for URL type that client never initiated. That's violation of the protocol.

- Nonce is used for server to keep track of which response the client is ACKing or NACKing. It is driven by server. Server generates a nonce and client should always echo what it receives.

Reply:
Removed the check. Client will just handle every responses, even if server violates the protocol. It doesn't affect client's behavior even if that happens, as client will not take any resources from responses until there are some watchers interested in some resources in the responses.